File: python/tests/test_manifest.py
# Copyright Institute for Automotive Engineering (ika), RWTH Aachen University
# SPDX-License-Identifier: Apache-2.0

from __future__ import annotations

from pathlib import Path

import pytest

from pcod_common import manifest


def _minimal_payload() -> dict:
    return {
        "schema_version": manifest.SCHEMA_VERSION,
        "artifact": {
            "bundle_name": "pbod_fp16_gpu_onnx_test",
            "export_format": "onnx_fp16_gpu",
            "backend": "onnx",
            "precision": "fp16",
            "device": "cuda",
            "files": {
                "model": "model.onnx",
                "checkpoint": "checkpoints/best.pt",
                "resolved_training_config": "config/resolved_training_config.yml",
            },
            "triton": {
                "enabled": False,
            },
            "inputs": [
                {
                    "name": "point_features",
                    "dtype": "float16",
                    "shape": ["batch", 100, 1],
                }
            ],
            "outputs": [
                {
                    "name": "reg_logits",
                    "dtype": "float16",
                    "shape": ["batch", 400, 7],
                }
            ],
        },
        "frozen_contract": {
            "preprocessing": {
                "max_num_points": 100,
                "num_point_features": 1,
                "point_cloud_range": {
                    "x": [-1.0, 1.0],
                    "y": [-1.0, 1.0],
                    "z": [-1.0, 1.0],
                },
                "voxel_size": {
                    "x": 0.1,
                    "y": 0.1,
                    "z": 0.1,
                },
                "point_feature_normalization": {
                    "type": "value_threshold",
                    "epsilon": 1e-6,
                },
            },
            "postprocessing": {
                "grid_size": {"x": 10, "y": 10},
                "num_classes": 1,
                "class_names": ["car"],
            },
            "model": {
                "stride": [2, 1, 2],
                "up_stride": [1, 1, 2],
                "first_up_stride": 1,
                "pillar_map_size": [10, 10],
                "pillar_map_range": [[-1.0, 1.0], [-1.0, 1.0], [-1.0, 1.0]],
            },
        },
        "runtime_defaults": {
            "preprocessing": {
                "point_feature": {
                    "value_threshold": 1.0,
                }
            },
            "postprocessing": {
                "class_score_threshold": 0.0,
                "nms": {
                    "score_threshold": 0.2,
                    "iou_threshold": 0.5,
                    "max_num_objects": 10,
                },
            },
        },
    }


def test_validate_manifest_accepts_minimal_payload():
    manifest.validate_manifest(_minimal_payload())


def test_load_manifest_reads_yaml(tmp_path: Path):
    path = tmp_path / "model_manifest.yml"
    path.write_text(
        """
schema_version: "2.0"
artifact:
  bundle_name: pbod
  export_format: onnx_fp32_gpu
  backend: onnx
  precision: fp32
  device: cuda
  files:
    model: model.onnx
    checkpoint: checkpoints/best.pt
    resolved_training_config: config/resolved_training_config.yml
  triton:
    enabled: false
  inputs:
    - name: point_features
      dtype: float32
      shape: [batch, 4, 1]
  outputs:
    - name: reg_logits
      dtype: float32
      shape: [batch, 16, 7]
frozen_contract:
  preprocessing:
    max_num_points: 4
    num_point_features: 1
    point_cloud_range:
      x: [-1.0, 1.0]
      y: [-1.0, 1.0]
      z: [-1.0, 1.0]
    voxel_size:
      x: 0.5
      y: 0.5
      z: 0.5
    point_feature_normalization:
      type: value_threshold
      epsilon: 1.0e-6
  postprocessing:
    grid_size:
      x: 2
      y: 2
    num_classes: 1
    class_names: [car]
  model:
    stride: [2]
    up_stride: [1]
    first_up_stride: 1
    pillar_map_size: [4, 4]
    pillar_map_range: [[-1.0, 1.0], [-1.0, 1.0], [-1.0, 1.0]]
runtime_defaults:
  preprocessing:
    point_feature:
      value_threshold: 12.0
  postprocessing:
    class_score_threshold: 0.1
    nms:
      score_threshold: [0.2]
      iou_threshold: 0.5
      max_num_objects: 16
"""
    )
    payload = manifest.load_manifest(path)
    assert payload["artifact"]["files"]["checkpoint"] == "checkpoints/best.pt"


def test_validate_manifest_rejects_wrong_schema_version():
    payload = _minimal_payload()
    payload["schema_version"] = "1.0"
    with pytest.raises(ValueError):
        manifest.validate_manifest(payload)


@pytest.mark.parametrize("missing_key", ["artifact", "frozen_contract", "runtime_defaults"])
def test_validate_manifest_rejects_missing_required_key(missing_key):
    payload = _minimal_payload()
    payload.pop(missing_key)
    with pytest.raises(ValueError, match="missing required key"):
        manifest.validate_manifest(payload)


def test_validate_manifest_requires_runtime_value_threshold_for_value_threshold_norm():
    payload = _minimal_payload()
    payload["runtime_defaults"]["preprocessing"]["point_feature"]["value_threshold"] = 0.0
    with pytest.raises(ValueError, match="value_threshold"):
        manifest.validate_manifest(payload)


def test_validate_manifest_rejects_frozen_value_threshold_key():
    payload = _minimal_payload()
    payload["frozen_contract"]["preprocessing"]["point_feature_normalization"]["value_threshold"] = 1.0
    with pytest.raises(ValueError, match="unsupported key"):
        manifest.validate_manifest(payload)


def test_validate_manifest_accepts_per_class_nms_thresholds():
    payload = _minimal_payload()
    payload["frozen_contract"]["postprocessing"]["num_classes"] = 2
    payload["frozen_contract"]["postprocessing"]["class_names"] = ["car", "pedestrian"]
    payload["runtime_defaults"]["postprocessing"]["nms"]["score_threshold"] = [0.1, 0.2]
    manifest.validate_manifest(payload)


@pytest.mark.parametrize(
    ("value", "expected"),
    [
        (None, []),
        (0.25, [0.25]),
        ([0.1, 0.2], [0.1, 0.2]),
    ],
)
def test_score_threshold_list(value, expected):
    assert manifest.score_threshold_list(value) == expected


def test_validate_manifest_rejects_unknown_artifact_key():
    payload = _minimal_payload()
    payload["artifact"]["hardware_compatible"] = True
    with pytest.raises(ValueError, match="unsupported key"):
        manifest.validate_manifest(payload)


def test_validate_manifest_rejects_bundle_escaping_file_path():
    payload = _minimal_payload()
    payload["artifact"]["files"]["checkpoint"] = "../best.pt"
    with pytest.raises(ValueError, match="bundle root"):
        manifest.validate_manifest(payload)

File: tests/test_manifest.cpp
// Copyright Institute for Automotive Engineering (ika), RWTH Aachen University
// SPDX-License-Identifier: Apache-2.0

#include "pcod_common/model_manifest.hpp"

#include <cassert>
#include <fstream>
#include <stdexcept>
#include <string>

/** Run manifest parsing and unknown-key validation checks. */
int main() {
  const std::string path = "./test_manifest.yml";
  std::ofstream out(path);
  out << "schema_version: '2.0'\n";
  out << "artifact:\n";
  out << "  bundle_name: 'pbod_fp16_gpu_onnx_test'\n";
  out << "  export_format: 'onnx_fp16_gpu'\n";
  out << "  backend: 'onnx'\n";
  out << "  precision: 'fp16'\n";
  out << "  device: 'cuda'\n";
  out << "  files:\n";
  out << "    model: 'model.onnx'\n";
  out << "    checkpoint: 'checkpoints/best.pt'\n";
  out << "    resolved_training_config: 'config/resolved_training_config.yml'\n";
  out << "  triton:\n";
  out << "    enabled: false\n";
  out << "  inputs:\n";
  out << "    - name: 'point_features'\n";
  out << "      dtype: 'float16'\n";
  out << "      shape: ['batch', 100, 1]\n";
  out << "  outputs:\n";
  out << "    - name: 'reg_logits'\n";
  out << "      dtype: 'float16'\n";
  out << "      shape: ['batch', 10, 7]\n";
  out << "frozen_contract:\n";
  out << "  preprocessing:\n";
  out << "    max_num_points: 100\n";
  out << "    num_point_features: 1\n";
  out << "    point_cloud_range:\n";
  out << "      x: [-1.0, 1.0]\n";
  out << "      y: [-1.0, 1.0]\n";
  out << "      z: [-1.0, 1.0]\n";
  out << "    voxel_size:\n";
  out << "      x: 0.1\n";
  out << "      y: 0.1\n";
  out << "      z: 0.1\n";
  out << "    point_feature_normalization:\n";
  out << "      type: value_threshold\n";
  out << "      epsilon: 1e-6\n";
  out << "  postprocessing:\n";
  out << "    grid_size:\n";
  out << "      x: 10\n";
  out << "      y: 10\n";
  out << "    num_classes: 1\n";
  out << "    class_names: ['car']\n";
  out << "  model:\n";
  out << "    stride: [2, 1, 2]\n";
  out << "    up_stride: [1, 1, 2]\n";
  out << "    first_up_stride: 1\n";
  out << "    pillar_map_size: [10, 10]\n";
  out << "    pillar_map_range: [[-1.0, 1.0], [-1.0, 1.0], [-1.0, 1.0]]\n";
  out << "runtime_defaults:\n";
  out << "  preprocessing:\n";
  out << "    point_feature:\n";
  out << "      value_threshold: 1.0\n";
  out << "  postprocessing:\n";
  out << "    class_score_threshold: 0.0\n";
  out << "    nms:\n";
  out << "      score_threshold: 0.2\n";
  out << "      iou_threshold: 0.5\n";
  out << "      max_num_objects: 10\n";
  out.close();

  auto manifest = pcod_common::LoadModelManifest(path);
  pcod_common::ValidateModelManifest(manifest);
  assert(manifest.frozen_contract.preprocessing.max_num_points == 100);
  assert(manifest.runtime_defaults.preprocessing.point_feature.value_threshold == 1.0f);

  const std::string invalid_path = "./test_manifest_invalid.yml";
  std::ofstream invalid(invalid_path);
  invalid << "schema_version: '2.0'\n";
  invalid << "artifact:\n";
  invalid << "  bundle_name: 'pbod_fp16_gpu_onnx_test'\n";
  invalid << "  export_format: 'onnx_fp16_gpu'\n";
  invalid << "  backend: 'onnx'\n";
  invalid << "  precision: 'fp16'\n";
  invalid << "  device: 'cuda'\n";
  invalid << "  hardware_compatible: true\n";
  invalid << "  files:\n";
  invalid << "    model: 'model.onnx'\n";
  invalid << "    checkpoint: 'checkpoints/best.pt'\n";
  invalid << "    resolved_training_config: 'config/resolved_training_config.yml'\n";
  invalid << "  triton:\n";
  invalid << "    enabled: false\n";
  invalid << "  inputs:\n";
  invalid << "    - name: 'point_features'\n";
  invalid << "      dtype: 'float16'\n";
  invalid << "      shape: ['batch', 100, 1]\n";
  invalid << "  outputs:\n";
  invalid << "    - name: 'reg_logits'\n";
  invalid << "      dtype: 'float16'\n";
  invalid << "      shape: ['batch', 10, 7]\n";
  invalid << "frozen_contract:\n";
  invalid << "  preprocessing:\n";
  invalid << "    max_num_points: 100\n";
  invalid << "    num_point_features: 1\n";
  invalid << "    point_cloud_range:\n";
  invalid << "      x: [-1.0, 1.0]\n";
  invalid << "      y: [-1.0, 1.0]\n";
  invalid << "      z: [-1.0, 1.0]\n";
  invalid << "    voxel_size:\n";
  invalid << "      x: 0.1\n";
  invalid << "      y: 0.1\n";
  invalid << "      z: 0.1\n";
  invalid << "    point_feature_normalization:\n";
  invalid << "      type: value_threshold\n";
  invalid << "      epsilon: 1e-6\n";
  invalid << "  postprocessing:\n";
  invalid << "    grid_size:\n";
  invalid << "      x: 10\n";
  invalid << "      y: 10\n";
  invalid << "    num_classes: 1\n";
  invalid << "    class_names: ['car']\n";
  invalid << "  model:\n";
  invalid << "    stride: [2, 1, 2]\n";
  invalid << "    up_stride: [1, 1, 2]\n";
  invalid << "    first_up_stride: 1\n";
  invalid << "    pillar_map_size: [10, 10]\n";
  invalid << "    pillar_map_range: [[-1.0, 1.0], [-1.0, 1.0], [-1.0, 1.0]]\n";
  invalid << "runtime_defaults:\n";
  invalid << "  preprocessing:\n";
  invalid << "    point_feature:\n";
  invalid << "      value_threshold: 1.0\n";
  invalid << "  postprocessing:\n";
  invalid << "    class_score_threshold: 0.0\n";
  invalid << "    nms:\n";
  invalid << "      score_threshold: 0.2\n";
  invalid << "      iou_threshold: 0.5\n";
  invalid << "      max_num_objects: 10\n";
  invalid.close();

  bool rejected_unknown_key = false;
  try {
    (void)pcod_common::LoadModelManifest(invalid_path);
  } catch (const std::runtime_error&) {
    rejected_unknown_key = true;
  }
  assert(rejected_unknown_key);
  return 0;
}
